src/app/controllers/sap.php

```diff
@@ -3,7 +3,22 @@
 class sap extends Controller {
 
     function index() {
+        $this->load_library('token_generator_lib', 'token_generator');
+        $csrf_token = $this->token_generator->getCSRFToken();
+
+        session_start();
+        if (isset($_SESSION['csrf_token'])) {
+            $oldToken = $_SESSION['csrf_token'];
+        } else {
+            $oldToken = false;
+        }
+        $_SESSION['csrf_token'] = $csrf_token;
+
         if ($_SERVER['REQUEST_METHOD'] == 'POST') {
+            if (isset($_POST['csrf_token']) && $oldToken != $_POST['csrf_token']) {
+                http_response_code(400);
+                die('Invalid request.');
```

Comment:
CSRF failure can also happen if session times out. So, not necessarily only when someone is trying to break things.